app/src/main/kotlin/com/wire/android/notification/CallNotificationManager.kt

```diff
@@ -62,75 +63,121 @@ class CallNotificationManager @Inject constructor(
+                            }
+                        }
+                    }.flatten()
+                }
+                .scan(emptyList<CallNotificationData>() to emptyList<CallNotificationData>()) { (previousCalls, _), currentCalls ->
+                    currentCalls to (currentCalls - previousCalls.toSet())
+                }
                 .distinctUntilChanged()
-                .reloadIfNeeded()
-                .collectLatest { incomingCallData ->
-                    if (incomingCallData == null) {
-                        hideIncomingCallNotification()
-                    } else {
-                        appLogger.i("$TAG: showing incoming call")
-                        showIncomingCallNotification(incomingCallData)
+                .flatMapLatest { (allCurrentCalls, newCalls) ->
+                    reloadCallNotification
+                        .map { (userIdString, conversationIdString) ->
+                            allCurrentCalls to allCurrentCalls.filter { // emit call that needs to be reloaded as newOrUpdated
+                                it.userId.toString() == userIdString && it.conversationId.toString() == conversationIdString
+                            }
+                        }
+                        .filter { (_, newCalls) -> newCalls.isNotEmpty() } // only emit if there is something to reload
+                        .onStart { emit(allCurrentCalls to newCalls) }
+                }
+                .collectLatest { (allCurrentCalls, newCalls) ->
+                    // remove outdated incoming call notifications
+                    hideOutdatedIncomingCallNotifications(allCurrentCalls)
+                    // show current incoming call notifications
+                    appLogger.i("$TAG: showing ${newCalls.size} new incoming calls (all incoming calls: ${allCurrentCalls.size})")
+                    newCalls.forEach { data ->
+                        showIncomingCallNotification(data)
                     }
                 }
         }
     }
 
-    fun reloadIfNeeded(data: CallNotificationData): Flow<CallNotificationData> = reloadCallNotification
-        .filter { reloadCallNotificationIds -> // check if the reload action is for the same call
-            reloadCallNotificationIds.userIdString == data.userId.toString()
-                    && reloadCallNotificationIds.conversationIdString == data.conversationId.toString()
+    @VisibleForTesting
+    internal fun hideOutdatedIncomingCallNotifications(currentIncomingCalls: List<CallNotificationData>) {
+        val activeNotifications = notificationManager.activeNotifications
+        val currentIncomingCallNotificationIds = currentIncomingCalls.map {
+            NotificationConstants.getIncomingCallId(it.userId.toString(), it.conversationId.toString())
+        }
+        activeNotifications.filter {
+            it.tag != null && it.tag.startsWith(INCOMING_CALL_ID_PREFIX) && !currentIncomingCallNotificationIds.contains(it.id)
```

Comment:
```suggestion
            it.tag?.startsWith(INCOMING_CALL_ID_PREFIX) == true && !currentIncomingCallNotificationIds.contains(it.id)
```

Reply:
For me my version feels to be more readable, this is subjective, I changed it anyway when implementing the Boris' suggestion